[Ready to review] Update rdm-spreadsheets.md (#1250)
alan-turing-institute/the-turing-way: book/website/reproducible-research/rdm/rdm-spreadsheets.md

As a simple rule, what can be exported in a text-only format, comma separated values (CSV) or tab separated values (TSV), can be considered as the data, other functions should be avoided when using these programm for research data. This includes: 
- changing font, color or borders,
- using functions, 
- merging cells (this one is particularly problematic), 
- using specific cell formats (especially dates, see below). 

As a test for your spreadsheet compatibility with reproducible research, export it in .csv format and reopen it: if you can still get all the information from that sheet, you are good.

```

> Tip: If you want to use color to help with a rapid highlight in your document, create a new column to enter that information (then it becomes is data) and use the conditional formatting tool to had color depending on the data entered. On top of the visual feedback, you can now also use that information to filter or sort your data and get faster to the highlighted cells.

Author: The conditional formatting tip is gone.

> Tip: If you want to use color to help with a rapid highlight in your document, create a new column to enter that information (then it becomes is data) and use the conditional formatting tool to had color depending on the data entered. On top of the visual feedback, you can now also use that information to filter or sort your data and get faster to the highlighted cells.

Reviewer: Sorry about that. The sentence was not very clear to me and I might have given up on rephrasing it.

### Spreadsheet design

In order to avoid mistakes and be most efficient, the same spreadsheet should be used to collect the data (even if this happens on a printed version of the spreadsheet) and analyse the data. The poses some design questions, especially for information which are unique for one experiment but that needs to be included in the data, as it changes between experiments (experimentator, temperature of the room,...). You indeed want that information into one column, but you would like to enter it only once (especially on a printed version). One solution is to move these columns on a second (Non-printed) page on the same spreadsheet, and playing with the headers and footers to enter the information on paper. 

> The way you enter the information (that is the way you design your headers and cell content) may be different depending on the analysis you want to perform later. One should still always try to be as generic and objective as possible, and think about additional analysis one may want to perform. As an example, let's suppose you are interested in a genetic tree of people with blue eyes. While you may have a column "has_blue_eyes" with the value true or false, it would be better to have a column about the eye colour (brown - blue- green) as it will be easy to create the  "has_blue_eyes" information from it, and allow additional analysis. (In this case, blue and green eyes might be the same gene for eyes, with a different skin color).

Author: I like the idea to use the iris dataset, but we would need an example of new data entries (otherwise we are not really using spreadsheet). I will think about how the dataset could have been collected in a bad way, and give the iris dataset as the good way. would this be better?

```
The way you enter the information (that is the way you design your headers and cell content) may be different depending on the analysis you want to perform later. One should still always try to be as generic and objective as possible, and think about additional analysis one may want to perform. 

As an example, let's suppose you are interested in depicting if the number of iris flowers whose sepal length is longer than 6 cm is different in three species. You may be inclined to record a column recording is-sepal-longer-than-6cm, but this will restrict the analysis you may perform. A better solution is to record the length of the sepal, and create automatically the categorization later.

If you are using R, you would then plot what you wanted with:
iris %>% dplyr::mutate ("is-sepal-longer-than-6cm" = ifelse(Sepal.Length >6, TRUE, FALSE)) %>%
  ggplot2::ggplot (aes (x=`is-sepal-longer-than-6cm` , fill= Species)) + ggplot2::geom_bar()

```

> The way you enter the information (that is the way you design your headers and cell content) may be different depending on the analysis you want to perform later. One should still always try to be as generic and objective as possible, and think about additional analysis one may want to perform. As an example, let's suppose you are interested in a genetic tree of people with blue eyes. While you may have a column "has_blue_eyes" with the value true or false, it would be better to have a column about the eye colour (brown - blue- green) as it will be easy to create the  "has_blue_eyes" information from it, and allow additional analysis. (In this case, blue and green eyes might be the same gene for eyes, with a different skin color).

Reviewer: I like this a lot!

If you are using excel in a team, be sure to all use the same version, as the default decimal value may change.

Finally, restricting values in specific columns (linking spreadsheets with taxonomies or ontologies) can be very useful. 
There are specific tools to help people fill such columns, as well as tools to validate the spreadsheets before integration in the analysis.

### Standard and versioning

A good spreadsheet design has instructive and intuitive header names, and it facilitates both data collection and analysis. 
Building such a spreadsheet design is difficult, takes time, iteration and consensus. It is therefore very useful to look for standard spreadsheet before designing your own, and share your design openly once you created yours. 
One should also use a version history of the spreadsheets (as they will evolve over time) and analysis script should mention the version number of the spreadsheet. 
Documentation of the spreadsheet, its version history, and the ontologies it is linked to, can be useful for future users. 

### Working in team: wrap-up

> If you are working with a team on data collection, make sure:
> - everyone is using the same software (and software version) to enter the data.
> - everyone is using the same version of the spreadsheet template.
> - everyone understand what each column represents, and the unit that should be used.
> - every column has a defined standard on how to enter data in it, or a taxonomy of terms that one can use.
> - One person is responsible to answer putative questions during data collection.
> - every spreadsheet is validated before entering the analysis workflow, and as soon as possible.

Reviewer: ```suggestion
If you are working with a team on data collection, make sure:
- everyone is using the same software (and software version) to enter the data.
- everyone is using the same version of the spreadsheet template.
- everyone understands what each column represents, and the unit that should be used.
- every column has a defined standard on how to enter data in it or taxonomy of terms that one can use.
- One person is responsible to answer putative questions during data collection.
- every spreadsheet is validated before entering the analysis workflow, and as soon as possible.
```

## Data manipulation and analysis

> **Do not do data manipulation or analysis in a spreadsheet program.**
> 
> In particular, only copy-paste from one spreadsheet to another if the process is used very rarely. 
> It is now very easy to read and combine the different spreadsheets in the analysis software, with the additional advantage that the software will return an error message if the headers do not fit.

Reviewer: ```suggestion
**Do not manipulate or analyse data in a spreadsheet program.**

In particular, only copy-paste from one spreadsheet to another if the process is used very rarely. 
It is now very easy to read and combine the different spreadsheets in the analysis software, with the additional advantage that the software will return an error message if the headers do not fit.
```

> 
> 
> 
> 

Reviewer: ```suggestion
```

## Summary

> While spreadsheet can be a very user friendly way to collect and share data, they can be the source of mistakes if used inappropriately. When aiming at doing reproducible analyses, one should design the spreadsheet for computer-readability and easy data analysis before starting data collection. Data manipulation and analysis in spreadsheets, in particular, is best avoided as it lead to non-reproducible workflows. Using version control and making the data read-only are two additional data management practices that can prevent accidents.

Reviewer: ```suggestion
While spreadsheets can be a very user-friendly way to collect and share data, they can also be the source of mistakes if used inappropriately. 
When aiming at developing a reproducible workflow for analyses, one should design the spreadsheet for both computer and human readability, and even before starting data collection, they should consider what makes their data analysis easy. 
Data manipulation and analysis in spreadsheets, in particular, is best avoided as it leads to non-reproducible workflows. 
Using version control and making the data read-only are two additional data management practices that can prevent accidents.
```

> Use a README texte file to explain naming conventions on top of the spreadsheet desing choices, so that it will also become clear to others what the file names mean. If you work in team, you should take particular care to the rules and make sure everyone is following them. 

Reviewer: ```suggestion
Use a README file to explain naming conventions and other design choices on top of the spreadsheet.
It will make it clear to others what the file names mean and what criteria to consider when designing an analysis workflow. 
If you work in a team, you should take particular care of the conventions and make sure everyone follows them. 
```

# Using spreadsheets for research data

Spreadsheets, such as Microsoft Excel files, googlesheets, and their Open Source alternative [(for instance) LibreOffice](https://www.libreoffice.org), are commonly used to collect, store, manipulate, analyse, and share research data.
Spreadsheets are convenient and easy-to-use tools for organising information into an easy to write and easy to read forms for humans. 
However, one should use them with caution, as the use of inappropriate spreadsheet is a major cause of mistakes in the data analysis workflow.
There is a collection of [horror-stories](http://www.eusprig.org/horror-stories.htm) that tells how the use of spreadsheets can ruin analysis-based studies due to unexpected behaviour of the spreadsheet or error-prone
editing processes.
Some of these mishaps are not unique to spreadsheets, but [many are](https://doi.org/10.1186/s13059-016-104).

Most problems can be avoided with the three rules explained below: 

> - use spreadsheet in a text only format (.csv or .tsv)
> - create tidy spreadsheets
> - and avoid manipulating and analysing data in spreadsheet software (this includes copy-paste). 

Author: ```suggestion
- use spreadsheet in a text only format (.csv or .tsv)
- create tidy spreadsheets
- make spreadsheets consistent (with each other) and implement rules for data entries.
- and avoid manipulating and analysing data in spreadsheet software (this includes copy-paste). 
```

There are very simple rules to facilitate data use, which go into the concept of [**tidy data**](https://en.wikipedia.org/w/index.php?title=Tidy_data&oldid=962241815). 
Furthermore, this specific format also allows for filtering and sorting data easily in spreadsheet software. 
In short:

- one colum = one variable (no more, no less, this implies that two header names can not be identical)
- one row = one sample
- one cell = one information
- **first row and only the first row is the header** 
- Headers name: do not include special character (including space), do not start with a number

| ![tidy table figure](../../figures/tidy-1.png)         |
| ------------------------------------------------------------------------------------ |
|Three rules which make a dataset tidy:|
|1. Each variable must have its own column.|
|2. Each observation must have its own row.|
|3. Each value must have its own cell.  |

There are data validation tools available, like https://goodtables.io, that allows you to check automatically whether your spreadsheets are tidy.

> 

Author: ```suggestion
## Consistent values

Especially if you work with several spreadsheets or if you work with a team during data collection, it is very important to make sure the same information will be entered with the same term, and a same term always convey the same information. 
If we take the example of iris data, if some people use different terms to record information for a specific column, such as naming the column `species` instead of `Species`, or using `iris setosa`, `set.` or `i.setosa` instead of `setosa`, the creation of a reproducible workflow  will be more difficult and errors may even be overlooked.  
Discrepancies often lead to errors, especially when the same terms could mean different things depending on who is entering the data. 
For example, indicating date as 02-03 will mean February the third in the USA, but March the second in Europe.

It is good practice to implement a `data dictionary`or a `taxonomy`of accepted terms and to document the convention used in a readme file.
Depending on the software you are using, you may be able to restrict the accepted values in specific columns.
If there is such a taxonomy or ontology available, using it may allow you (or others) to use the data in conjonction with other datasets.
For example, use the generic `male` and `female` terms for the sex of animal (without capitals, and without using abbreviation), as many ontologies use these terms.
In addition, you may want to use some extra tools to validate the spreadsheets before its integration in the analysis.

In addition, you should have clear rules about missing data points, using `NA`, `NULL` and/or empty cells is not trivial and may have different meanings (impossible data point, not recorded, or lost data point). 
Let's imagine, one would record the time spent before seeing the first pollinator landing on the iris flower measured previously, and that no pollinator was seen during the 5 minutes experiment.
If one reports `600` (the duration of the experiment in seconds), there will be no way to distinguish a case where no pollinator was seen, and one when one was seen at the end of the experiment (and you may forget that rule and treat 600 as a normal value).
If `NA` is reported, one may interpret this value as a non-existing data point (the experiment had not been performed).
An elegant solution is to have a second column stating whether a pollinator was seen during the experiment, where `TRUE`, `FALSE` and `NA` values are accepted.

Finally, you should also be aware of the default behaviour of your spreadsheet program, as it may be different for different programs, and for different versions of the same program.
For instance, the decimal is usually indicated with a comma in the French or German version of excel, but while a dot in enlish versions where the comma has no meaning
(`9,000`  will be translated into 9000 or 9 depending on the version you are using).
```

> 

Author: this should resolve the consistency comment you had!

> 

Reviewer: Thank you so much. This does make it very clear.

```
  

Headers names should be chosen with care, and when it is not perfectly clear what is meant and what unit is used, you may want to add some explanation in an external document. 
You may also share a sample spreadsheet to a colleague to receive feedback on how understandable your sheet is.

An alternative way is to put some explanations on the top of the sheet in the first rows before the headers.
By keeping human-readable information at the top of the file, one can better understand the data that starts in a row with the header.
This information can also help in analysing that data making sure that the scripts ignore the explanation lines and considers the explanation during analysis.
However, a good file with tidy columns and rows should not need extra explanation.

As for file name, the size of the headers is not an issue for computers, but for human readability it might be better to keep it short (up to 32 characters).

You do not have to think about the order of the columns for the analysis, as it has no importance for data analysis software. 
You can therefore completely optimise that parameter for the data collection step.

> ### Accepted values
> 
> Especially if you work with a team on data collection, it is very important to make sure everyone will enter the same information with the same term. 
> If we the example of iris data, if some people use different terms to record information for a specific column, such as naming the column "species" instead of "Species", or using "iris setosa", "set." or "i.setosa" instead of "setosa", the creation of a reproducible workflow  will be more difficult and errors may even be overlooked.  
> Discrepancies often lead to errors, especially when the same terms could mean different things depending on who is entering the data. 
> For example, indicating date as 02-03 will mean February the third in the USA, but March the second in Europe.
> 
> **Critical: make sure you use dots for decimal.** 
> If you are working with data and researchers in countries that follow a different convention, clearly documented those caveats and maintain all your spreadsheets using the same convention. 
> This is particularly important in several European countries (like Germany) where commas are normally used for decimal.
> If you are using excel in a team, be sure to use the same convention, else the default decimal value may end up varying from user to user.
> 
> Finally, restricting values in specific columns (linking spreadsheets with taxonomies or ontologies) can be very useful. 
> There are specific tools to help people fill such columns, as well as tools to validate the spreadsheets before integration in the analysis.

Author: ```suggestion

```

editing processes.
Some of these mishaps are not unique to spreadsheets, but [many are](https://doi.org/10.1186/s13059-016-104).

Most problems can be avoided with the three rules explained below: 
- use spreadsheet in a text only format (.csv or .tsv)

> - create tidy spreadsheets

Reviewer: ```suggestion
- create tidy and consistent spreadsheets
```

Author: What do you mean exactly with "consistent" ? This would need to be explained in the detailed description.

Author: Does it refer to different spreadsheet being consistent one with the others? (same unit used, same header names,...)

Author: google gives this in the first page : http://www.ssrb.org/resources/standards/ce/sma3/sheet_consistency, which is full of very, very bad advices...

Reviewer: Gees, consistent in the sense - if you have used a certain delimiter, stick with it in your data set. If you have a certain placeholder for missing data, stick with it. (and so on). What would that be called (if not consistent)? Any suggestion?

Author: delimiters is indeed included in "using .csv format". The other elements are, I think, included in the "Accepted values" section. Do you think we should had some information there, and do you think this should go into the main 3 rules (it would then be 4) ?

Author: I will close this without commit and work the text later

If the spreadsheet is poorly organised then it may be [difficult for collaborators](https://luisdva.github.io/pls-don't-do-this/) to easily [read-in and re-use](#FAIR) your
data for further analysis.

> The number one nightmare of data scientists is indeed a researcher coming with a question related to "a couple of spreadsheets to analyse". This is mainly because reading the data will be a very long process. There are very simple rules to facilitate data use, which go into the concept of [**tidy data**](https://en.wikipedia.org/w/index.php?title=Tidy_data&oldid=962241815). By the way, this specific format also allows for filtering and sorting data easily in spreadsheet software. In short:

Reviewer: ```suggestion
The number one nightmare of data scientists is indeed a researcher coming with a question related to "a couple of spreadsheets to analyse". 
This is mainly because reading the data will be a very long process. 
There are very simple rules to facilitate data use, which go into the concept of [**tidy data**](https://en.wikipedia.org/w/index.php?title=Tidy_data&oldid=962241815). 
Furthermore, this specific format also allows for filtering and sorting data easily in spreadsheet software. 
In short:
```

Reviewer: New sentence in the new line to make future review easier (see the style guide)

> You do not have to think about the order of the columns for the analysis, as it has no importance for data analysis software. You can therefore completely optimise that parameter for the data collection step.

Reviewer: ```suggestion
You do not have to think about the order of the columns for the analysis, as it has no importance for data analysis software. 
You can therefore completely optimise that parameter for the data collection step.
```

### Accepted values

> Especially if you work with a team on data collection, it is very important to make sure everyone will enter the same information with the same term. If we take the example above, if some people use "bluish", or "b" instead of blue, the analysis of the data will be more difficult. Discrpencies might lead to errors, expecially if the same terms could mean different things depending on who is entering the data. Dates are again a very good example as 02-03 will mean February the third in the USA, but March the second in Europe.

Reviewer: ```suggestion
Especially if you work with a team on data collection, it is very important to make sure everyone will enter the same information with the same term. 
If we take the example above, if some people use "bluish", or "b" instead of blue, the analysis of the data will be more difficult. 
Discrepancies might lead to errors, especially if the same terms could mean different things depending on who is entering the data. 
Dates are again a very good example as 02-03 will mean February the third in the USA, but March the second in Europe.
```

> **Critical: make sure you use dots for decimal.** or make sure to document that you used commas, that all spreadsheets are using the same convention. This is particulary important in Europe where commas are normally used for decimal, meaning .csv export will need quotes for each number cell to be importable. If you are using excel in a team, be sure to all use the same version, as the default decimal value may change.

Reviewer: ```suggestion
**Critical: make sure you use dots for decimal.** or make sure to document that you used commas, that all spreadsheets are using the same convention. 
This is particularly important in Europe where commas are normally used for decimal, meaning .csv export will need quotes for each numbered cell to be importance. 
If you are using excel in a team, be sure to all use the same version, as the default decimal value may change.
```

> Finally, restricting values in specific columns (linking spreadsheets with taxonomies or ontologies) can be very useful. There are specific tools to help people fill such columns, as well as tools to validate the spreadsheets before integration in the analysis.

Reviewer: ```suggestion
Finally, restricting values in specific columns (linking spreadsheets with taxonomies or ontologies) can be very useful. 
There are specific tools to help people fill such columns, as well as tools to validate the spreadsheets before integration in the analysis.
```

> A good spreadsheet design has instructive and intuitive header names, and it facilitate both data collection and analysis. Building such a spreadsheet design is difficult, takes time, iteration and consensus. It is therefore very useful to look for standard spreadsheet before designing your own, and share your design openly once you created yours. One should also use a version history of the spreadsheets (as they will evolve over time) and analysis script should mention the version number of the spreadsheet. A documentation of the spreadsheet, its version history, and the ontologies it is linked to, can be useful for future users. 

Reviewer: ```suggestion
A good spreadsheet design has instructive and intuitive header names, and it facilitates both data collection and analysis. 
Building such a spreadsheet design is difficult, takes time, iteration and consensus. It is therefore very useful to look for standard spreadsheet before designing your own, and share your design openly once you created yours. 
One should also use a version history of the spreadsheets (as they will evolve over time) and analysis script should mention the version number of the spreadsheet. 
Documentation of the spreadsheet, its version history, and the ontologies it is linked to, can be useful for future users. 
```

### Working in team: wrap-up

If you are working with a team on data collection, make sure:
- everyone is using the same software (and software version) to enter the data.
- everyone is using the same version of the spreadsheet template.
- everyone understand what each column represents, and the unit that should be used.
- every column has a defined standard on how to enter data in it, or a taxonomy of terms that one can use.
- One person is responsible to answer putative questions during data collection.
- every spreadsheet is validated before entering the analysis workflow, and as soon as possible.

## Data manipulation and analysis

**Do not do data manipulation or analysis in a spreadsheet program.**

> In particular, only copy-paste from one spreadsheet to another if the process is used very rarely. It is now very easy to read and combine the different spreadsheets in the analysis software, with the additional advantage that the software will return an error message if the headers do not fit.

Reviewer: ```suggestion
In particular, only copy-paste from one spreadsheet to another if the process is used very rarely. 
It is now very easy to read and combine the different spreadsheets in the analysis software, with the additional advantage that the software will return an error message if the headers do not fit.
```
